DataConnectors/AWS-CloudTrail-Ingestion-Lambda/IngestCloudTrailEventsToSentinel.ps1

```diff
@@ -0,0 +1,345 @@
+			$bits = [math]::Round(($eventsjson.length) / 20MB) + 1
+			$TotalRecords = $groupEvents[$d].Count
+			$RecSetSize = [math]::Round($TotalRecords / $bits) + 1
+			$start = 0
+			For ($x = 0; $x -lt $bits; $X++) {
+				IF ( ($start + $recsetsize) -gt $TotalRecords) {
+					$finish = $totalRecords
+				}
+				ELSE {
+					$finish = $start + $RecSetSize
+				}
+				$body = Convertto-Json ($groupEvents[$d][$start..$finish]) -Depth 5 -Compress
+				$result = Invoke-LogAnalyticsData -CustomerId $workspaceId -SharedKey $workspaceKey -Body $body -LogTable $Table -TimeStampField 'eventTime' -ResourceId $ResourceID                
+				if ($result -eq 200)
+				{
+					Write-Host "CloudTrail Logs successfully ingested to LogAnalytics Workspace under Custom Logs --> Table: $Table"
+				}
+				$start = $Finish + 1
+			}
+			$null = Remove-variable -name body        
+		}
+		Else {
+			#$logEvents = Convertto-Json $events -depth 20 -compress
+			$result = Invoke-LogAnalyticsData -CustomerId $workspaceId -SharedKey $workspaceKey -Body $eventsJson -LogTable $Table -TimeStampField 'eventTime' -ResourceId $ResourceID
+			if ($result -eq 200)
+			{
+				Write-Host "CloudTrail Logs successfully ingested to LogAnalytics Workspace under Custom Logs --> Table: $Table"
+			}
+		}
+	}
+	
+}
+
+foreach ($snsRecord in $LambdaInput.Records)
+{
+    $snsMessage = ConvertFrom-Json -InputObject $snsRecord.Sns.Message
+    foreach ($s3Event in $snsMessage.Records)
+    {
+        $s3BucketName = $s3Event.s3.bucket.name
+        $s3BucketKey = $s3Event.s3.object.key
+
+        Write-Host "Processing event for: bucket = $s3BucketName, key = $s3BucketKey"
+	
+		IF ($Null -ne $s3BucketName -and $Null -ne $s3BucketKey) {
+			$s3KeyPath = $s3BucketKey -Replace ('%3A', ':')			
+			$fileNameSplit = $s3KeyPath.split('/')			
+			$fileSplits = $fileNameSplit.Length - 1			
+			$fileName = $filenameSplit[$fileSplits].replace(':', '_')			
+			$downloadedFile = Read-S3Object -BucketName $s3BucketName -Key $s3BucketKey -File "/tmp/$filename"			
+			Write-Host "Object $s3BucketKey is $($downloadedFile.Size) bytes; Extension is $($downloadedFile.Extension)"
+			
+			IF ($downloadedFile.Extension -eq '.gz' ) {
+				$infile = "/tmp/$filename"				
+				$outfile = "/tmp/" + $filename -replace ($downloadedFile.Extension, '')					
+				Expand-GZipFile $infile.Trim() $outfile.Trim()
+				$null = Remove-Item -Path $infile -Force -Recurse -ErrorAction Ignore
+				$filename = $filename -replace ($downloadedFile.Extension, '')
+				$filename = $filename.Trim()
+			
+    
+				$logEvents = Get-Content -Raw -LiteralPath ("/tmp/$filename" ) 
+				$logEvents = $LogEvents.Substring(0, ($LogEvents.length) - 1)
+				$LogEvents = $LogEvents -Replace ('{"Records":', '')
+				$loglength = $logEvents.Length    
+				$logevents = Convertfrom-json $LogEvents -AsHashTable
+				$groupevents = @{}
+				$coreEvents = @()
+				$eventSources = @()
+				Foreach ($log in $logevents) {
```

Comment:
worth adding some comments to the this log parsing function, feel it is hard to read and a lot special character replacement, not sure what it does.